### DISCUSSION TITLE
is there a plan to support slot ?

### DISCUSSION THREAD

**Discussion by gicque** [vuejs/petite-vue#89] [Q&A]
slots can be usefull to compose components and are more usefull in some than props/attributes
Is there a plan to support slots and named slots ?

**Comment by XLordalX** [comment]
I second this.

**Comment by justJokee** [comment]
need this so much😭

**Comment by nervusdm** [comment]
This would be great

**Comment by hellogerard** [comment]
My hacky solution:

Call like this:

```js
<div v-scope="LayoutComponent()" @vue:mounted="mounted($el)">
  <slot class="hidden">
    <p>Slot contents</p>
  </slot>
</div>
```

Component:

```js
<script>
  const LayoutComponent = () => ({
    $template: '#layout-template',
    slot: null,
    mounted(el) {
      this.slot = el.querySelector('slot').innerHTML;
    },
  });
</script>

<template id="layout-template">
  <div class="container">
    <h2>My Layout</h2>
    <div v-html="slot"></div>
  </div>
</template>
```

**Comment by jmaximusix** [comment]
I also came up with a solution using nested templates using v-scope:

```html
<template id="sharedtemplate">
    <h1>Stuff all instances of the component share...</h1>
    <slot v-scope="FillSlot(slot)"></slot>
</template>

<div v-scope="MyComponent('#first')">
    <template id="first">
        <p>This is the slot content for the first component</p>
    </template>
</div>

<div v-scope="MyComponent('#another')">
    <template id="another">
        <p>This is the slot content of the second</p>
    </template>
</div>

```

```js
import { createApp } from 'https://unpkg.com/petite-vue?module';

function MyComponent(slot) {
    return {
        $template: '#sharedtemplate',
        slot
    }
}

function FillSlot(slot) {
    return {
        $template: slot
    }
}

createApp({
    MyComponent,
    FillSlot,
}).mount("#app")
```

The benefit of this solution is that you can easily define other Components with multiple slots to be filled (and all can use the same "FillSlot()" function still), the disadvantage is you always have to use ids which could get annoying when creating many instances of your component.
So in my opinion this solution is somewhere between hacky and elegant, but I'll let you be the judge

**Reply by jikkuatwork** [reply]
Very interesting! I wonder whether this approach can do multiple components?

**Reply by jmaximusix** [reply]
Yes, absolutely, you can do multiple different components. I should also note that the template html tag I nested (for better cohesion/readability) can technically be put anywhere in the html.
However I still consider this more of a workaround than a solution. It works fine, I added a few components in my project and it's still better than having to repeat the whole thing, but I wouldn't recommend doing something complicated with lot's of components and instances like that.
Would be awesome if it were properly supported tho!

**Reply by jikkuatwork** [reply]
True, I wish they supported it. I would have happily accepted the speed/size trade off for DX.

**Reply by jikkuatwork** [reply]
Also, can `slot` can have components with state/functions right?

I am having trouble wiring up components with state & function; if its not too much to ask, can you look at this [Pen](https://codepen.io/jikkuatwork/pen/dyLBKVb?editors=1010)?

Will your solution be able to slot the `Counter({ initialValue })` inside the `Panel({ title })` while maintaining the individual state of the Counter?
Also, does your solution enable providing arguments for Counter & Panel while doing it?

**Reply by jikkuatwork** [reply]
Just got it working: [petite-vue/component-nesting-example](https://codepen.io/jikkuatwork/pen/GRLbGer)
